Disable os_reload function

jianqiu opened the pull request:
As os_reload behavior might prevent open source teams from really using this CPI, bosh-softlayer-cpi should honor some property from the CPI configuration or from the resource pool cloud properties which controls OS reload.  There are 2 approaches to disable os_reload function:
1. Indicate DisableOsReload flag in resource pool cloud properties:

```
resource_pools:
- cloud_properties:
    Bosh_ip: <%=director_ip%>
    Datacenter:
      Name: <%=data_center_name%>
    EphemeralDiskSize: 50
    HourlyBillingFlag: true
    LocalDiskFlag: true
    MaxMemory: 8192
    StartCpus: 4
    VmNamePrefix: <%=bluemix_env_name%>-core-
    DisableOsReload: true
  name: coreNode
  network: default
  size: 14
  stemcell:
    name: <%=stemcell%>
    version: <%=stemcell_version%>
```
1. Indicate DisableOsReload flag in cpi.json of softlayer_cpi job:

```
{
  "cloud": {
    "plugin": "softlayer",
    "properties": {
      "softlayer": {
        "username": "******",
        "apiKey": "******"
      },
      disableOsReload: true,
      "agent": {
        "ntp": [],
        "blobstore": {
          "provider": "dav",
          "options": {
            "endpoint": "http://127.0.0.1:25250",
            "user": "agent",
            "password": "agent"
          }
        },
        "mbus": "nats://nats:nats@127.0.0.1:4222"
      }
    }
  }
}
```

jianqiu commented:
@maximilien 
This PR is ready for your review, thanks!

mattcui commented:
@maximilien Please help review when you get time, better to close this PR before @jianqiu leaves. Thanks.

maximilien commented on api/dispatcher/json.go:
```diff
@@ -179,6 +179,6 @@ func (c JSON) localDiskFlagNotSet(reqString string) {
 	} else {
 		bslcommon.LocalDiskFlagNotSet = true
-	//	c.logger.DebugWithDetails(jsonLogTag, "localDiskFlag is not specified in the input json, will set true as the default value", "")
+		//	c.logger.DebugWithDetails(jsonLogTag, "localDiskFlag is not specified in the input json, will set true as the default value", "")
```
Why are we keeping this comment?

maximilien commented:
Looks OK, though couple things I do not like:

1. I wish we would have feature flag structure so that OS reload is one of these. Similar to BOSH agent. This would be since as the CPI is used more and more, I anticipate additional feature flags to be added. Thoughts?

2. You have no tests for this. I want some unit tests at a minimal.

3. Address my minor comment.

Best,

Max

jianqiu commented:
@maximilien  @mattcui 

Hi Max,

I had addressed all your comments, and it is ready for your review now, thanks!

For comment 1, the there is one struct defined to containing all feature flags for SoftLayer

```
type SoftLayerConfig struct {
	Username        string `json:"username"`
	ApiKey          string `json:"apiKey"`
	DisableOsReload bool   `json:"disableOsReload,omitempty"`
}
```

it works similiar to options settings in bosh-agent.

For comment 2, updated with new unit test cases to cover this feature flag.

For comment 3. done.

jianqiu commented:
@maximilien  @mattcui 

I had moved disableOsReload flag to a combined struct which will contain all feature flags for softlayer. Please review my codes, thanks!

```
type SoftLayerConfig struct {
	Username       string                `json:"username"`
	ApiKey         string                `json:"apiKey"`
	FeatureOptions bslcvm.FeatureOptions `json:"featureOptions"`
}
```

maximilien commented on softlayer/vm/softlayer_virtual_guest_creator.go:
```diff
@@ -36,17 +38,23 @@ func NewSoftLayerCreator(vmFinder Finder, softLayerClient sl.Client, agentOption
 		switch network.Type {
 		case "dynamic":
-			if len(network.IP) == 0 {
+			if cloudProps.DisableOsReload || c.featureOptions.DisableOsReload {
```
Why check both? Should it we just set `c.featureOptions.DisableOsReload` if `cloudProps.DisableOsReload`

maximilien commented:
OK merging. Thanks for addressing previous comments. I added one question. Please answer and address as a short PR if needed. Thanks 👍
